libfwupd/fwupd-client.c

```diff
@@ -4240,6 +4250,54 @@ fwupd_client_upload_bytes_finish (FwupdClient *self, GAsyncResult *res, GError *
 	return g_task_propagate_pointer (G_TASK(res), error);
 }
 
+/* this is bad; we dlopen libsoup-2.4.so.1 and get the gtype manually
+ * to avoid deps on both libcurl and libsoup whilst preserving ABI */
+static void
+fwupd_client_ensure_soup_session (FwupdClient *self)
+{
+	FwupdClientObjectNewFunc func = NULL;
+	FwupdClientPrivate *priv = GET_PRIVATE (self);
+	GType soup_gtype;
+
+	/* already set up */
+	if (priv->soup_session != NULL)
+		return;
+
+	/* known GType, just create */
+	soup_gtype = g_type_from_name ("SoupSession");
+	if (soup_gtype != 0) {
+		priv->soup_session = g_object_new (soup_gtype, NULL);
+		return;
+	}
+
+	/* load the library at rumtime, leaking the module */
+	if (priv->soup_module == NULL) {
+		g_autofree gchar *fn = NULL;
+		const gchar *libdirs[] = {
+			FWUPD_LIBDIR,
+			"/usr/lib64",
+			"/usr/lib",
+			NULL };
+		for (guint i = 0; libdirs[i] != NULL && priv->soup_module == NULL; i++) {
+			fn = g_build_filename (libdirs[i], "libsoup-2.4.so.1", NULL);
```

Comment:
hardcoding this extension won't work in !linux

Reply:
Valid, but I don't think it matters; this is just for gnome-software.

Reply:
Maybe don't even bother compiling the code for mingw builds?

Reply:
Yes, I'd feel a lot happier if all this evil was behind some kind of ifdef. Will fix.